[ENH] Move to pyproject.toml

Moritz-Alexander-Kern opened the pull request:
This is a long one!

# Original PR Body by Moritz
This PR moves Elephant from setup.py to pyproject.toml, aligning with PEP 518.

Key changes include:

- Introduction of `pyproject.toml`
- `setup.py` is still needed to compile the C-code for the fim module.

The reason is that we need OS dependend compile args which are dynamically defined in `setup.py`.
The `pyproject.toml` does not natively support conditional logic like os_name == 'windows' directly within the configuration. The TOML format is static and doesn't allow for dynamic evaluation of conditions based on the operating system.

To achieve platform-specific compile arguments, we have to use a Python script, such as a `setup.py`, which can programmatically determine the OS and apply the correct compile arguments.

In this context, as a configuration file for setuptools, `setup.py` is not deprecated, see also: https://packaging.python.org/en/latest/discussions/setup-py-deprecated/

---
# Changes

## pyproject.toml
- Created `pyproject.toml` with full project metadata (name, version, description, authors, license,
  URLs, classifiers)
- Moved to the SPDX licence expression format per PEP 639; removed the legacy `License ::` classifier string
- Switched to auto-discovery of packages via `setuptools.find_packages()`, this correctly picks up
subpackages such as `elephant.asset`
- Added `include-package-data = true` so non-Python data files are included in the wheel install
- Added classifiers for Python 3.13 and 3.14
- Removed the long-unused `six` dependency
- Added `viziphant` to the `tutorials` optional-dependency group
- Dependencies are now declared via [dynamic metadata](https://setuptools.pypa.io/en/latest/userguide/pyproject_config.html#dynamic-metadata), pointing back to the `requirements/*.txt` files. This keeps a single source of truth for dependencies

## CI modernisation (`.github/workflows/CI.yml`)
- **Merged pip test jobs**: Three separate pip-test jobs (Ubuntu, macOS, Windows) merged into a single job with `matrix.include`, eliminating duplication
- **Composite action**: Added a composite action(`.github/actions/restore-elephant-data`) to avoid code duplication of restore action that restored cached elephant-dataset and is shared by the pip, MPI and conda jobs
- Enabled `enableCrossOsArchive: true` on the data cache so a cache built on one OS can be restored on
another, this was added to build to both restore action and cache build action (`.github/workflows/cache_elephant_data.yml`)
- Bumped `actions/cache/`, `actions/setup-python` and `actions/checkout` to newer versions to fix issue with Node.js 20 actions being deprecated
- Made python version common accross jobs
- **Per-job coveralls uploads + finish job**: every job now uploads its own partial coverage report to coveralls, and a new `finish` job combines them at the end. This is to fix the earlier race condition where last coverall job was being usede to add PR comment.
- Added `pytest-cov` to `requirements/requirements-tests.txt` to avoid duplicate install steps for each job
- **Fast CI mode**: added a `fast` workflow-dispatch boolean input that limits the matrix to a subset of jobs
- Cleaned up workflow triggers 
- Cleaned up the DOC BUILD and DOC TEST jobs

## Environment files
- Refactored `requirements/environment*.yml` conda environment files to import from the corresponding `requirements/*.txt` files instead of duplicating the dependency list (system-level conda-only packages are still listed directly) 

## Bug fix for Windows path
- `elephant/datasets.py`: switched to `local_file.as_posix()` when constructing the `ValueError` message in `download_datasets`. On Windows, `pathlib` rendered backslashes, causing `test_download_with_cache_dir_with_failed_integrity_check` to fail because the assertion looks for the POSIX-style `repo_path` substring.

Commit:
added metadata to pyproject.toml

Commit:
added dependencies

Commit:
added build-system

Commit:
fix formatting

Commit:
build fim module using setup.py

Commit:
remove description from CITATION.txt, leave reference to docs

Commit:
fix formatting of pyproject.toml

Commit:
use requirement files for dependencies

Commit:
add somesy config

Commit:
update somesy config

Commit:
remove somesy config

Commit:
update license field

Moritz-Alexander-Kern reviewed (commented):
build wheels

Commit:
add wheel to build dependencies

Moritz-Alexander-Kern reviewed (commented):
build wheels

Commit:
define packages for setuptools

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
Merge branch 'NeuralEnsemble:master' into enh/pyproject.toml

Moritz-Alexander-Kern committed:
Merge branch 'NeuralEnsemble:master' into enh/pyproject.toml

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
Merge branch 'master' into enh/pyproject.toml

Moritz-Alexander-Kern committed:
remove setup.py

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update ci build wheel action

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update build wheel action

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update ci build wheel action

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update cibuildwheel action

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update os for building wheels

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
change compile args for linux

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update python version for cibuildwheels

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update ci buildwheel action

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update compile args

Moritz-Alexander-Kern committed:
re add setup.py

Moritz-Alexander-Kern committed:
update cibuildwheel version

Moritz-Alexander-Kern committed:
typo

Moritz-Alexander-Kern committed:
typo

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
build wheels

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update python version

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern committed:
update cibuildwheel

Moritz-Alexander-Kern reviewed (commented):
build wheels

Moritz-Alexander-Kern commented:
Editable install with `setup.py` will be deprecated with pip 25.0

> DEPRECATION: Legacy editable install of elephant==1.2.0b1 from file:/// (setup.py develop) is deprecated. pip 25.0 will enforce this behaviour change. A possible replacement is to add a pyproject.toml or enable --use-pep517, and use setuptools >= 64. If the resulting installation is not behaving as expected, try using --config-settings editable_mode=compat. Please consult the setuptools documentation for more information. Discussion can be found at https://github.com/pypa/pip/issues/11457

Moritz-Alexander-Kern committed:
Merge branch 'master' into enh/pyproject.toml

Moritz-Alexander-Kern committed:
Merge branch 'master' into enh/pyproject.toml

Moritz-Alexander-Kern reviewed (commented):
build wheels

CozySocksAlways commented:
- [ ] Update requirements file to remove upper bound matplot lib (notebooks were okay as well as sphinx docs).
- [x] Add viziphant to tutorials requirements.

CozySocksAlways committed:
Merge branch 'master' into enh/pyproject.toml

CozySocksAlways committed:
Removed wheels as build requirement. Wheels was historically listed, no longer needed according to setuptools doc pages, unless imported in setup.py

CozySocksAlways committed:
Replaced requirements.txt with dependencies in pyproject.toml, making
pyproject.toml file the source of truth for dependencies

CozySocksAlways commented on setup.py:
```diff
@@ -30,7 +15,7 @@
             '-DMODULE_NAME=fim', '-DUSE_OPENMP', '-DWITH_SIG_TERM',
             '-Dfim_EXPORTS', '-fopenmp', '/std:c++17'],
-        optional=True
+        #optional=True
```
Why was this commented out? 
- Was added in  #494 to overcome the compile fail issue, but didn't work at the time with development (editable) install?
- Commenting it out will cause `optional` defaulting to `None` which is Falsy. So install fails on build?

CozySocksAlways committed:
moved to new SPDX licence expression specified by PEP 639, removed legacy classifier string for licence

CozySocksAlways committed:
Removed newline in desc as per warning from setuptools that newlines are not allowed in summary

CozySocksAlways committed:
replaced hardcoded package location with auto discovery using setuptools.find, this picks up (includes) subpackages ike asset etc. into wheels

CozySocksAlways committed:
Added option to include non py data in wheel using include-package-data

CozySocksAlways reviewed (commented):
trigger build wheels ci

CozySocksAlways committed:
Reset option=True to continue install in case compilation fails

CozySocksAlways committed:
Added classifiers for newer python versions; Removed unused dependency six, was removed in an old commit 2333599

CozySocksAlways committed:
Cleaned up triggers for CI.yml

CozySocksAlways committed:
Combined pip install and test for all OS jobs in CI.yml, relying on git bash as shell for windows

CozySocksAlways committed:
Merge branch 'master' into enh/pyproject.toml

CozySocksAlways committed:
Bumbed cache action restore to v5; removed redundant lines that set env vars

CozySocksAlways committed:
Updated MPI job to use newer python version; Removed coverage for MPI job, as there is a race condition now with pip coverage (can later combine them with a job that combines coveralls reports); Moved dependency to toml file

CozySocksAlways committed:
Replacing dataset restore steps in PIP job with a composite action that can be reused accross jobs

CozySocksAlways committed:
Temporarily keeping only one python version for pip job

CozySocksAlways committed:
Added viziphant to tutorials requirement

CozySocksAlways committed:
Enable cross OS caching on dataset caching workflow

CozySocksAlways committed:
Added method to render error message in posix standard, so that test_download_with_cache_dir_with_failed_integrity_check catches it on windows as well

CozySocksAlways committed:
Replacing dataset restore steps in MPI job with composite action

CozySocksAlways committed:
Replaced dataset restore in CONDA job with composite action; Also removed coveralls for CONDA job as it is runs almost same tests as MPI or PIP jobs

CozySocksAlways committed:
Cleaned up DOC BUILD test job

CozySocksAlways committed:
Cleaned up DOC TEST job

CozySocksAlways committed:
Reverting 540e20f, decided to use dynamic metadata for dependencies and optional dependencies, this avoids need of dedicated CI to check and maintain requirements files

CozySocksAlways committed:
Added a fast CI option, i.e. run a subset of job matrix

CozySocksAlways committed:
Made environment.yml files also depend on requirement files, except for system packages

CozySocksAlways committed:
Merge remote branch 'NeuralEnsemble:master' into HEAD

CozySocksAlways committed:
Added fix for RTD failure

CozySocksAlways committed:
Updated CI jobs to each upload a coveralls report;
Added a finish job to combine coveralls report;
Added pytest-cov pluggin to req-test to avoid duplicated
install statements in CI file;

CozySocksAlways committed:
Bumbed github action versions based on deprication warnings on worflow due to Node js issue

CozySocksAlways commented:
Postponing addition of GPU tests with group runner available on jugit.fz-juelich.de to a different PR.

CozySocksAlways commented:
Once PR approved, full CI should be triggered. (Run once for each PR approval)

mdenker commented on pyproject.toml:
```diff
@@ -0,0 +1,56 @@
+        "Topic :: Scientific/Engineering"]
+keywords = ["neuroscience", "neurophysiology", "electrophysiology", "statistics", "data-analysis"]
+readme = {file = "README.md", content-type = "text/markdown"}
+requires-python = ">=3.8"
```
I feel like this line is giving me problems when doing the following:
- Create a new environment using `uv venv`
- Doing a `uv pip install .` in the elephant directory
- Calling `uv run python`m which gives me an error about dependencies not being satisfyable:uv responds that while I am currently using Python 3.14, the `reuires-python` indicates >=3.8. This is however incompatible with numpy 2. Therefore, uv fails, as it cannot resolve for any of the support python versions. I do not fully understand this logic, however, I guess we should drop 3.8 despite this being a hot fix. Or maybe I am missing something?

CozySocksAlways replied:
Yup I could recreate this issue, it's an error from my side. The setup.py on our latest release has a requirement on `python>=3.9`, so I'll fix this now on the toml file. 

As a side note, the latest release of numpy (2.4.6) has a requirement of `python>=3.11`, though lower version of numpy 2.0 still support python 3.9. So we should definitely fix this on our next minor release.

mdenker reviewed (changes requested):
Thank you for this extensive PR, besides the comment about Python versions, it looks good to me so far -- I guess we need to see how the remaining CI jobs work out. I tested installs on Linux and Windows, which seem to work (including the fim module unter linux). However, the Python 3.8 issue that prevents installation of the package in a uv managed pip environment is something we should address.

CozySocksAlways committed:
Added clause to setup job to filter out only approved cases for pull_request_review trigger; Made all other jobs dependent on setup job, so they are also skipped based on setup job

CozySocksAlways commented:
Test duplication in case of pull_request_review trigger (i.e. both pull_request_review triggered tests pollutes the existing PR checks summary made with the checks triggered on PR synchronize) is the right behavior on GitHub and cannot be avoided as GitHub treats these two as separate events. Refer https://github.com/orgs/community/discussions/26940

As some form of relief, an if clause has been added on the setup job that makes it only run when pull_request_review is approved and not when a review comment is added.

CozySocksAlways committed:
Fixed required python version based on review comments; This was an typo when migrating from setup.py

CozySocksAlways committed:
Removed classifier for python 3.8; Added python 3.9 to full test suite

kohlerca reviewed (approved):
The PR looks good to me. I tested locally in Ubuntu, using `mamba` and `uv`, and the installation succeeds.

CozySocksAlways commented:
- [x] Update names of tests for consistency (i.e with MPI and conda)

CozySocksAlways committed:
Updated names of CI tests for brevity and consistency
